stockroom/repository.py

```diff
@@ -1,9 +1,82 @@
+from typing import Union
 from pathlib import Path
 from hangar import Repository
-from .utils import get_stock_root
+from .utils import get_stock_root, get_current_head, set_current_head
 
 
-def init(name, email, overwrite):
+class StockRepository:
+    """
+    A StockRoom wrapper class for hangar repo operations. Every hangar repo
+    interactions that is being done through stockroom (other than stock init)
+    should go through this class. Unlike hangar Repository, this class constructor
+    assumes the hangar repo is already initialized. Three class variables keep the
+    reference required for all the storages to interact with hangar repository.
+        1. Hangar repo object
+        2. Hangar repository root
+        3. Write checkout
+    Since these are class variables, they are not re-created on each initialization
+    and hence we can make sure at most, there will be only one writer checkout object
+    active at any point in time (Hangar will make sure this assertion is True anyways)
+    """
+    _write_checkout = None
+    _root: Union[Path, None] = None
+    _hangar_repo = None
+
+    def __init__(self):
+        self._setup()
+
+    @classmethod
+    def _setup(cls):
+        """
+        Setup the stock repository object by assuming the current working directory as
+        the point of operation. It doesn't require user to pass the path but used the
+        ``cwd``. This function creates the hangar repo object only if it is not creates
+        already. It also assumes that the hangar repo is already initialized and expect
+        the presence of stock file
+        """
+        if cls._hangar_repo is None:
+            cwd = Path.cwd()
+            cls._root = get_stock_root(cwd)
+            if not cls._root:
+                raise FileNotFoundError("Could not find stock file. "
+                                        "Did you forget to `stock init`?")
+            cls._hangar_repo = Repository(cls._root)
+            if not cls._hangar_repo.initialized:
+                raise RuntimeError("Repository has not been initialized")
+
+    @classmethod
+    def _teardown(cls):
+        cls._hangar_repo._env._close_environments()
+        cls._write_checkout = None
+        cls._root = None
+        cls._hangar_repo = None
+
+    @classmethod
+    def checkout(cls, write=False):
```

Comment:
Similar to the `_setup` method above, these don't seem like they need to be `classmethod`'s, but are actuall just actions attached to an attribute of the class? What's the rationale here?

Reply:
same as above

Reply:
see https://github.com/tensorwerk/stockroom/pull/1#discussion_r361956040